docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+}
+```
+
+Our `AdderMonitor` prints out the results of the `Adder` each cycle and
+signals an error if the `Adder` returns an incorrect result. It has two
+`AdderMonitorNode`s for taking in addends from the `AdderDriver`, and one
+`AdderMonitorNode` to take in the sum from the `Adder`.
+
+```scala mdoc
+/** monitor (sink) */
+class AdderMonitor(width:Int)(implicit p: Parameters) extends LazyModule {
+  val nodeA   = new AdderMonitorNode(UpwardParam(width))
+  val nodeB   = new AdderMonitorNode(UpwardParam(width))
+  val nodeSum = new AdderMonitorNode(UpwardParam(width))
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val error = Output(Bool())
+    })
+
+    // print output
+    printf(p"${nodeA.in.head._1} + ${nodeB.in.head._1} = ${nodeSum.in.head._1}")
+
+    // basic correctness checking
+    io.error := nodeSum.in.head._1 =/= nodeA.in.head._1 + nodeB.in.head._1
+  }
+
+  override lazy val desiredName = "AdderMonitor"
+}
+```
+
+### Creating the Top
+
+Our top-level module will be a test bench that instantiate our `Adder` and its
+peripherals. Notice that different `width` parameters are passed into Diplomacy
+via `driver` and `checker`. As described in our `AdderNodeImp`, we expect the
+smaller of the two widths to actually be used.
+
+In this module, we also bind our nodes. Sinks are on the left-hand side, while
+sources are on the right.
+
+```scala mdoc
+/** top-level connector */
+class AdderTestHarness()(implicit p: Parameters) extends LazyModule {
+  val adder = LazyModule(new Adder)
+  val driver = LazyModule(new AdderDriver(8))    // passing in downward-traveling width to Diplomacy
+  val checker = LazyModule(new AdderMonitor(4))  // passing in upward-traveling width
+
+  // node connections
+  adder.node := driver.nodeA
+  adder.node := driver.nodeB
+
+  checker.nodeA := driver.nodeA
+  checker.nodeB := driver.nodeB
+  checker.nodeSum := adder.node
+
+  lazy val module = new LazyModuleImp(this) {
+    when (checker.module.io.error) {
+      printf("something went wrong")
+    }
+  }
+
+  override lazy val desiredName = "AdderTestHarness"
+}
+```
+
+We have used the most typical node binding here, but other types of bindings
+can be found in [Nodes.scala](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/Nodes.scala).
+
+### The Generated Verilog
+
+Now we are ready to generate the Verilog for our circuit. The `DummyConfig` is
+empty, since we don't need any parameters specified through a `Config`.
+
+```scala mdoc:silent
+/** empty config */
+class DummyConfig extends Config((_,_,_) => { case _ if false => "" })
+
+val verilog = (new ChiselStage).emitVerilog(
+  LazyModule(new AdderTestHarness()(new DummyConfig)).module
+)
+```
+
+Below is the generated Verilog for our modules! Note the parameterized ports
+in `Adder`, `AdderDriver`, and `AdderChecker` all get the lower width (8).
+The LFSR is also parameterized to 8 bits.
```

Comment:
Wasn't 4 the smaller width?

Reply:
oops yes, I shortened the widths in the code because the generated LSFRs were taking up too much space :P